M/orders/products component

RaimondasSapola opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop / 1.7.6.x / 1.6.1.x
| Description?  | Please be specific when describing the PR. <br> Every detail helps: versions, browser/server configuration, specific module/theme, etc.
| Type?         | bug fix / improvement / new feature / refacto / critical
| Category?     | FO / BO / CO / IN / WS / TE
| BC breaks?    | Does it break backward compatibility? yes/no
| Deprecations? | Does it deprecate an existing feature? yes/no
| Fixed ticket? | [Issue](https://github.com/PrestaShop/PrestaShop/issues), please write "Fixes #[issue number]" here.
| How to test?  | Please indicate how to best verify that this PR is correct.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

jusurb committed:
Fixes issue Undefined index _route

not sure why _route is expected to be part of caller parameters but it's not, the route is actually in the caller_route parameter, so we first merge parameters with defaults array to make sure _route key is set. + will not overwrite params with the same name

RaimondasSapola committed:
Product component handlers and query results

RaimondasSapola committed:
Updated handler php doc

RaimondasSapola committed:
order product component assets

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create-order-map.js:
```diff
@@ -71,4 +71,26 @@ export default {
+  customizedFieldInputWrapper: '.js-customized-field-input-wrapper',
+  customizedFieldTypes: {
+    0: '#cartProductCustomizedFieldInputTemplate-0',
+    1: '#cartProductCustomizedFieldInputTemplate-1',
```
Why is it hardcoded?

RaimondasSapola replied:
It is because indexes represent types of customization fields. Im not sure how to make it dynamic for this instance.

sarjon replied:
Would it be possible to use `const`?

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+const $ = window.$;
+
+/**
+ * Page Object for "Create order" page
```
I think docs are incorrect.

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+  _handleProductSearch(event) {
+    const name = $(event.target).val();
+
+    if (name.length < 3) {
```
No magic numbers, define it as const instead. :)

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+   * @private
+   */
+  _handleProductSearch(event) {
+    const name = $(event.target).val();
```
Could you extract `$(event.target)` into separate var? Something like `$searchInput` since this selector is used multiple times.

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+      return;
+    }
+
+    $.ajax($(event.target).data('url'), {
```
Data attribute could be more specific, maybe `product-search-url`, wdyt?

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+      },
+    }).then((response) => {
+      this.products = JSON.parse(response);
+      this._renderProductSearchResult();
```
I'd prefer to pass `products` as `_renderProductSearchResult` arguments instead.

RaimondasSapola replied:
i would as well but because i need products data in other function it needs to be set globally here or in _renderProductSearchResult method. do you think setting in _renderProductSearchResult is better?

sarjon replied:
Ohh, then it doesn't matter I guess.

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+      }
+
+      if (typeof response.responseJSON !== 'undefined') {
+        showErrorMessage(response.responseJSON.message);
```
What if `message` does not exist?

RaimondasSapola replied:
can it be removed completely? Because i copied this if statement mindlessly and now im not sure if its needed at all :D

sarjon replied:
Not sure, maybe it can.

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+  _renderProductSearchResult() {
+    $(createOrderPageMap.productSelect).empty();
+
+    for (const [index, value] of this.products.entries()) {
```
What's `value`? Could you improve naming? Maybe it's a `product`?

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+        name += ' - ' + value.formatted_price;
+      }
+
+      if (value.combinations === null && index === 0) {
```
Would be nice to extract these into more meaningful variables.

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+   * @param product
+   * @private
+   */
+  _fillRelatedProductFields(product) {
```
I think name `relatedProduct` could be misinterpreted.

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+   * @private
+   */
+  _handleProductChange(event) {
+    const index = $(event.target).find(':selected').data('index');
```
It's better to use `currentTarget` instead of `target`.

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+      }
+
+      const name = combination.attribute + ' - ' + combination.formatted_price;
+      $(createOrderPageMap.combinationsSelect).append($('<option></option>')
```
It would be nice not to repeat same selectors. :)

sarjon commented on admin-dev/themes/new-theme/js/pages/order/create/order-product-component.js:
```diff
@@ -0,0 +1,313 @@
+    }
+
+    $(createOrderPageMap.combinationsSelect)
+      .on('change', (event) => this._handleCombinationChange(event));
```
It is necessary to add event listener every time?

RaimondasSapola replied:
It is because combination select is not ind DOM if there is no combinations so after rendering it i have to set listener every time. Is there maybe a better way?

sarjon replied:
I think you need to delegate event using `$parent.on('change', '#selector', callback)`

sarjon commented on src/Core/Domain/Order/Query/SearchProductsForOrderCreation.php:
```diff
@@ -0,0 +1,51 @@
+     */
+    public function __construct(string $phrase)
+    {
+        $this->phrase = $phrase;
```
Does it have any constraints? Can it be empty?

sarjon commented on src/Core/Domain/Order/QueryResult/ProductCustomizationFields.php:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * @param ProductCustomizationField $productCustomizationField
+     */
+    private function addProductCustomizationField(ProductCustomizationField $productCustomizationField): void
```
Private methods should be defined at the end of file.

sarjon commented on src/Core/Domain/Order/QueryResult/ProductForOrderCreationCombinations.php:
```diff
@@ -0,0 +1,61 @@
+
+namespace PrestaShop\PrestaShop\Core\Domain\Order\QueryResult;
+
+class ProductForOrderCreationCombinations
```
Maybe `ProductCombinationsForOrderCreation`?

sarjon commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -362,6 +364,35 @@ public function addPaymentAction(int $orderId, Request $request): RedirectRespon
+            $productsForOrderCreation = $this->getQueryBus()->handle(new SearchProductsForOrderCreation($searchPhrase));
+
+            if (empty($productsForOrderCreation->getProducts())) {
+                return new Response('', Response::HTTP_NOT_FOUND);
```
I'm don't think it's correct use of `not found` status. When searching, it's ok to return `200` with empty collection.

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/orders.yml:
```diff
@@ -79,3 +80,9 @@ admin_orders_add_payment:
+  path: /search
+  methods: [GET]
+  defaults:
+    _controller: PrestaShopBundle:Admin/Product:searchProducts
```
What is this route?

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/orders.yml:
```diff
@@ -11,6 +11,7 @@ admin_orders_create:
   defaults:
     _controller: PrestaShopBundle:Admin\Sell\Order\Order:create
     _legacy_controller: AdminOrders
+    _legacy_link: AdminOrders:addorder
```
Does it mean you enable new page already?

RaimondasSapola replied:
It is still WIP 😅

sarjon commented on src/PrestaShopBundle/Resources/config/services/adapter/order.yml:
```diff
@@ -115,3 +115,14 @@ services:
+        - '@prestashop.adapter.legacy.context'
+        - '@prestashop.core.localization.locale.repository'
+        - "@=service('prestashop.adapter.legacy.context').getContext().language.getLocale()"
+
```
Empty line.

sarjon commented on src/PrestaShopBundle/Resources/config/services/bundle/serializer.yml:
```diff
@@ -0,0 +1,17 @@
+      - '@serializer.name_converter.camel_case_to_snake_case'
+
+  # Snake case serializer definition
+  prestashop.bundle.snake_case_serializer_json:
```
Maybe it would be nice to have tests for it?

RaimondasSapola replied:
It is just a specifically configured symfony serializer. Is it really necessary?

sarjon replied:
Hmmmm, not sure, maybe not if it's covered by symfony itself.

zuk3975 replied:
Im not sure its worth introducing snake case serializer at all.
Symfony controllers already have $this->json() method for json response. As JSON doesn't really have any strict convention, it seems like an overhead to me.

zuk3975 replied:
or maybe it does make sense after this - https://github.com/PrestaShop/PrestaShop/issues/15977#issuecomment-542780827

sarjon commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/Create/cart.html.twig:
```diff
@@ -88,11 +90,12 @@
             <tbody>
               <tr>
-                <td><img src="http://prestashop.local/2-small_default/hummingbird-printed-t-shirt.jpg" alt=""></td>
+                <td><img src="http://presta.local/img/p/2/2-small_default.jpg" alt=""></td>
```
It should not be left hardcoded. :wink:

jusurb committed:
changed [] to array() for more BC

jusurb committed:
fixed BC break with php 5.6 syntax

RaimondasSapola committed:
Minor CR fixes

jusurb committed:
Update src/PrestaShopBundle/Controller/Admin/CommonController.php

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

RaimondasSapola committed:
Add product to cart action

RaimondasSapola committed:
Add product to cart submission assets

ashishsharawat committed:
in.xml

ashishsharawat committed:
Update localization/in.xml

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

ashishsharawat committed:
Update localization/in.xml

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

ashishsharawat committed:
Update localization/in.xml

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

ashishsharawat committed:
Update localization/in.xml

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

sarjon committed:
Add date formatting for order status

sarjon committed:
Add generating invoice & delivery slip from order view page

sarjon committed:
Fix view for virtual orders

RaimondasSapola committed:
add product handling

RaimondasSapola committed:
Add product frontend handling to submit product data

zuk3975 committed:
rename html IDs to kebab-case. Fix missing argument in getErrorMessage. Restrict GET method in addProductToCartAction

zuk3975 committed:
Implements products list rendering after adding to cart

L3RAZ committed:
Changed the context variable passed to bindSwapChange(context)

The context variable is used inside bindSwapSave(context) function to find the selected values.

Previously the variable passed to the function was referring to the button inside the .click() function.

Now new variable "swap_container" is passed which refers to the found swap container.

atomiix committed:
fix number of DL & expiration date being displayed in the download virtual product mail

atomiix committed:
Display placeholder data instead of variables in download_product mail preview

zuk3975 committed:
Rename product searcher and move to product controller. WIP refactoring js

zuk3975 committed:
Refactoring products block js

202-ecommerce committed:
Fix notice since console include config.inc.php

See also #15515 
https://github.com/PrestaShop/PrestaShop/pull/15515#issuecomment-536637885

202-ecommerce committed:
remove require config.inc.php on module cli

202-ecommerce committed:
remove config.inc.php on theme command

202-ecommerce committed:
remove config.inc.php on generate mail command line

202-ecommerce committed:
remove config.inc.php on sql update command line

202-ecommerce committed:
remove config.inc.php on configuration command

zuk3975 committed:
Switch js products and combination to object with id instead of array

zuk3975 committed:
refactoring customization fields. Added todo

zuk3975 committed:
mvp of customization fields adding

zuk3975 committed:
Moves cartrules rendering parts to renderer

Prestaworks committed:
Update AdminImagesController.php

zuk3975 committed:
Remove unused methods

zuk3975 committed:
implement events on addProduct action

zuk3975 committed:
Adds customer renderer.

zuk3975 committed:
Trigger cart info update after product is added to cart.

zuk3975 committed:
Fix private method

zuk3975 committed:
Modifies api route names

atomiix committed:
Add missing placeholder in the Private Note textarea

zuk3975 committed:
Adds remove product from cart action

zuk3975 committed:
Initialize event listeners instead of calling them multiple times

atomiix committed:
Delete supplier's address only if no other shops use it

Commit:
Add demo file for brands and addresses

Commit:
Add faker For Brand and Address

Commit:
Add new eslint rule

PierreRambaud committed:
Merge pull request #16114 from atomiix/add-missing-private-note-placeholder

Add missing placeholder in the Private Note textarea

Commit:
Update function select by visible text for time saving

Commit:
Add selectors for BO Menu

Commit:
Change value of puppeteer slowMo

Commit:
Add function to delete file

Commit:
Add Brand Pages

Commit:
Add functional tests for brands and addresses

atomiix committed:
Change checkout button wording

Commit:
Update tests

Commit:
delete update of slowMo

Commit:
Fix name of class

Commit:
Using clickAndWaitForNavigation function

zuk3975 committed:
refactoring

Commit:
Update test

Commit:
adding common test for going into brands page

matks committed:
Merge pull request #16126 from atomiix/change-checkout-button-wording

Change checkout button wording

matks committed:
Merge pull request #16022 from sarjon/m/order/actions

Migrate order action and status block

matks committed:
Merge pull request #16072 from 202-ecommerce/patch-13

Remove require config.inc.php on all command and fix a notice

matks committed:
Merge pull request #15906 from Gamesh/patch-2

Fixes issue Undefined index _route

matks committed:
Merge pull request #16045 from L3RAZ/admin-swap-select-fix

Swap select value submision fix in admin forms

Commit:
Fix common Test goToBrandsPage

Commit:
Rename common test file

matks committed:
Merge pull request #16123 from atomiix/fix-supplier-edit-multistore

Delete supplier's address only if no other shops use it

Commit:
Update reset function to include returning nomber of elements in list

Commit:
Adding empty line between its

Commit:
Fix call common Test BOBase

Commit:
Fix name Called for BoBase Common Tests

Commit:
Fix name of commonTest

Commit:
Delete common tests gotoBrands page

Commit:
Update name of catalog Menu link in BO

SimonGrn committed:
Merge pull request #16156 from boubkerbribri/updateLink

Tests - Update name of catalog Menu link in BO

matks committed:
Merge pull request #16097 from Prestaworks/develop

Improvement on product image generation for high definition images.

Commit:
Merge remote-tracking branch 'upstream/develop' into functionalBrands

Progi1984 committed:
Merge pull request #16000 from ashishsharawat/patch-1

Update Indian tax rates

Progi1984 committed:
Merge pull request #16049 from atomiix/fix-virutal-product-links-in-mail-template

Fix number of DL & expiration date being displayed in the download virtual product mail

matthieu-rolland committed:
Merge pull request #16125 from boubkerbribri/functionalBrands

Functional Tests - Add BO tests for Brands

zuk3975 committed:
Implement delay on search and request abort. Implement addCustomizationFieldsHandler return customizationId

zuk3975 committed:
add const for customization field types
